x-pack/plugins/rule_registry/server/event_log/log/event_query.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { estypes } from '@elastic/elasticsearch';
+import { set } from '@elastic/safer-lodash-set';
+import { Logger } from 'kibana/server';
+import { IIndexReader } from '../elasticsearch';
+import { truthy } from '../utils/predicates';
+import { IEventQuery } from './public_api';
+
+export interface EventQueryParams {
+  indexReader: IIndexReader;
+  logger: Logger;
+  baseRequest: estypes.SearchRequest;
+}
+
+export class EventQuery<TEvent> implements IEventQuery<TEvent> {
+  private readonly indexReader: IIndexReader;
+  private readonly baseRequest: estypes.SearchRequest;
+
+  constructor(params: EventQueryParams) {
+    this.indexReader = params.indexReader;
+    this.baseRequest = params.baseRequest;
+  }
+
+  public async execute(): Promise<TEvent[]> {
+    const response = await this.indexReader.search<TEvent>(this.baseRequest);
+    return response.body.hits.hits.map((hit) => hit._source).filter(truthy);
+  }
+
+  public async search<TDocument = TEvent>(
+    extraRequest: estypes.SearchRequest
+  ): Promise<estypes.SearchResponse<TDocument>> {
+    const finalRequest = mergeRequests(this.baseRequest, extraRequest);
+    const response = await this.indexReader.search<TDocument>(finalRequest);
+    return response.body;
+  }
+}
+
+const mergeRequests = (
+  base: estypes.SearchRequest,
+  ext: estypes.SearchRequest
+): estypes.SearchRequest => {
+  const result: estypes.SearchRequest = {
+    ...base,
+    ...ext,
+    body: {
+      ...base.body,
+      ...ext.body,
+    },
+  };
+
+  const baseBool = base.body?.query?.bool;
+  const extBool = ext.body?.query?.bool;
+
+  if (baseBool != null || extBool != null) {
+    const mergedBool: estypes.BoolQuery = {
+      filter: mergeBoolClauses(baseBool?.filter, extBool?.filter),
+      must: mergeBoolClauses(baseBool?.must, extBool?.must),
+      must_not: mergeBoolClauses(baseBool?.must_not, extBool?.must_not),
+      should: mergeBoolClauses(baseBool?.should, extBool?.should),
+      minimum_should_match: extBool?.minimum_should_match ?? baseBool?.minimum_should_match,
+      boost: extBool?.boost ?? baseBool?.boost,
+      _name: extBool?._name ?? baseBool?._name,
```

Comment:
I think this query merge strategy may lead to some unintuitive results with the merged queries. In addition to the good points @xcrzx brought up, I would expect the logical result of merging two queries, A and B, to be `A && B`, i.e. returning the subset of documents that would be returned by both queries if executed independently. However, by merging the `should` arrays here, it's possible for the query to return documents that would not have been returned by the base query alone.

Boolean query logic details:
Given a query A, with the structure
```
filter: A_1,
must: A_2,
must_not: A_3,
should: A_4
```
the returned docs will satisfy `A_1 && A_2 && !A_3 && A_4`. Each sub clause A_1...4 can be an array, and the internal logic of A_1...3 is essentially `A_n = A_n_1 && A_n_2 ...` (although for the `must_not` array, each element is logically negated).

A_4, the `should` array, is different though - `A_4 = A_4_1 || A_4_2 ...`. Effectively, for query A to evaluate to true, at least one of the elements in the `should` array must be true.

When we merge the `should` arrays here, we make a new array for the merged query `C_4 = A_4_1 || B_4_1 || A_4_2 || B_4_2 ...`. Thus for this merged query to evaluate to true, `A_4` does *not* have to be true (B_4 can be true instead to satisfy the merged query), and thus A does not have to be true.

If we do indeed want the overall logic to be `C = A && B` then perhaps it would be simplest to nest A and B inside another `filter` clause? e.g. something like `result.body.query.bool = {filter: [base.body.query, ext.body.query]}`. This would apply the `ext` query as sort of a secondary filter that is applied after the base filter to further refine the results.

Reply:
@marshallmain you are absolutely right, thank you for your hard look and finding this issue 🙏 

I could fix it like you suggested - by nesting the two queries in a parent filter: `result.body.query.bool = {filter: [base.body.query, ext.body.query]}`. That would be good enough for me personally. On the other hand, I dig what @xcrzx suggested in https://github.com/elastic/kibana/pull/98353#discussion_r631279942 and I think separating the two read APIs makes probably even more sense.

@marshallmain @xcrzx your opinion?

Reply:
I would suggest separating APIs as this is the easiest option for now. That way, we can keep the initial implementation as lightweight as possible. Later, as we collect more usage statistics, we can adjust the API and implement merge strategies based on common usage patterns if needed. But for now, I would prefer to keep the API as simple as possible.

Reply:
Addressed